Refactor attempt

milansky opened the pull request:
I wrote down some questions within the code as comments.
When you have the time please take a look :) 

Can't wait to discuss new ideas and features to implement here :)

Commit:
Refactor attempt 1

Very poor attempt to create a drawing class and "engine" class to create a window with resolution and FoV being gathered/entered via the class constructor

Commit:
Delete engine.cpp

Had this extra file created, don't know how that happened.

Commit:
Questions

Adding some questions as comments

petrSchreiber commented on src/glfw-sandbox/drawings.h:
```diff
@@ -0,0 +1,88 @@
+#include <math.h>
+
+
+class drawObjects
```
Tip: What is the casing convention for C++ classes? Does this adhere to it? :)

milansky replied:
CamelCase? like "DrawObjects"?

petrSchreiber replied:
The current code has camelCase, which means first letter small.

I think for classes it is more common to use PascalCase, which is that DrawObjects you suggest :)

...and yes,

petrSchreiber commented on src/glfw-sandbox/drawings.h:
```diff
@@ -0,0 +1,88 @@
+
+public:
+
+	void triangle()
```
Tip: Do the drawing methods require any instance data? If not, how can we decorate them?

milansky replied:
When you think about it, no, they don't need to be instanced.

But maybe calling them via classes without instancing first?
Not sure what do you mean by decorating them :(

petrSchreiber replied:
In Python, when method does not use instance data, you decorate it with @staticmethod.

I would just suggest to do C++ analogy here :)

milansky replied:
Well... templates comes to mind but not sure if they are the right fit there.

For calling functions without instancing we could use DrawObjects::triangle()

petrSchreiber replied:
`DrawObjects::triangle();` is perfectly fine :)

petrSchreiber commented on src/glfw-sandbox/drawings.h:
```diff
@@ -0,0 +1,88 @@
+#pragma once
```
General tip: In C++, I think Nico called the header files .hpp - to differentiate from .h digestible by C.

General tip: .hpp file should contain just the interface, accompanying .cpp should contain the implementation then. Do you remember how we achieved that?

Xelios26 replied:
I was generally also confused with this, because I saw numerous times .h files accompanied with .cpp
I believe you just use something like a class definition in .hpp file
e.g my_awesome_class.hpp
```
{
    class my_awesome_class
    {
    public:
        void do_something();
    };

}
```
then in implement_my_awesome_class.cpp

```
#include "my_awesome_class.hpp" 
#include <iostream> 

void my_awesome_class::do_something()
{
    cout << "Yes I am doing something!" << endl;
}
```

Again I am not sure :)

petrSchreiber replied:
Exactly!!!

I discussed this with Hanka at work and she mentions that this allows to keep the hpp files free of unnecessary includes, among other things - like, you can quickly overview the interface.

petrSchreiber commented on src/glfw-sandbox/engine.h:
```diff
@@ -0,0 +1,96 @@
+#pragma once
```
// Same tip with hpp and cpp applies here

milansky replied:
Honestly that popped up from the Visual Studio by default, we can use header guards as Nico explained :)

petrSchreiber replied:
No no, `#pragma once` is nice, my comment means I would expect to have `engine.hpp` and `engine.cpp` :)

petrSchreiber commented on src/glfw-sandbox/engine.h:
```diff
@@ -0,0 +1,96 @@
+
+	}
+
+	void opengl_intialize() {
```
Do we need to call this explicitly, or can we do it for the user?

milansky replied:
We could add/move this to the constructor and have it called once its initialized?

petrSchreiber replied:
Or, we could make it a private method, which gets called in the constructor or before we do first action with drawing.

milansky replied:
I figured this out, now I understand what you mean but there is a problem.

If its initialized before glfwMakeContextCurrent(window); there is a bug and there is unexpected behavior.
For example the grid is not rendered and if you attempt to use the keyboard arrows for control you're gone...

Not sure how to explain this... its very strange.
Right now I'm trying to figure out a way to resolve this without messing out the code too much.

milansky replied:
In other words the opengl_intialize() needs to be called after glfwMakeContextCurrent

petrSchreiber replied:
Yes, exactly - not sooner

petrSchreiber commented on src/glfw-sandbox/main.cpp:
```diff
@@ -9,136 +9,31 @@
-        std::cout << "Failed to initialize GLFW library" << std::endl;
-        return 1;
-    }
+    drawObjects drawx;
```
I really like how stuff became simpler here 🥇

petrSchreiber reviewed (commented):
Few comments on the code, I am really happy you started the refactoring 🤩

petrSchreiber commented on src/glfw-sandbox/engine.h:
```diff
@@ -0,0 +1,96 @@
+	int resWidth{};
+	int resHeight{};
+	
+	// Could you please clarify what these three do?
+	// Are these ponters? If posssible just a little clarification how this works
+	// This might be tied to main.cpp line 23-32
+	char *title;
+	GLFWmonitor *monitor;
+	GLFWwindow *share;
```
Yes, these are pointers to the respecitve types. GLFW is C, not C++, so this is why we have this here :)

milansky replied:
Thank you :)

petrSchreiber commented on src/glfw-sandbox/engine.h:
```diff
@@ -0,0 +1,96 @@
+		if (!glfwInit())
+		{
+			std::cout << "Failed to initialize GLFW library" << std::endl;
+			//return 1; Why I can't return 1 here?
```
...because this is a constructor. You can, still raise an exception, in theory, on the other side, it is not recommended to fire exceptions from constructor. We are trapped! :)

petrSchreiber commented on src/glfw-sandbox/main.cpp:
```diff
@@ -9,136 +9,31 @@
+    // What is the difference between this and the previous "GLFWwindow* window;" call ???
+    /*
+    
+    struct window {
```
Struct is something like a class, but from the ancient times :) It helps you to pack data together.

petrSchreiber replied:
`GLFWwindow *window` is just a declaration.

`struct window {...};` means "I would like to keep everything related to window on one place, and have width, height, title, monitor and ... glfwwindow pointer packed in one struct (=something like a class, will explain on call).

petrSchreiber commented on src/glfw-sandbox/main.cpp:
```diff
@@ -171,7 +67,7 @@ int main() {
         // Rendering will go here
         
-        triangleZ -= 0.001f;
+        triangleZ -= 0.001f; //What is the difference between 0.001f and 0.001? Don't know how to google this question :( 
```
If C++ does not require us to write f, we maybe don't need to. It simply indicates the number should be treated as float.

milansky replied:
I've been reading some stuff on StackOverflow but they are really old. Maybe we should ask Hana?

petrSchreiber replied:
I know the answer! The result might surprise you :D Stay tuned for the afternoon session :)

Commit:
Change file names and static function

Changed the .h files to .hpp

Added static to functions in drawing.
Don't know how to implement engine to call opengl initialize in private.

Commit:
Minor cleanup

Minor cleanup and adding some notes

Commit:
Controls and cleanup

Added two classes for controls.

One can be instanced and the other static functions.
Also some minor cleanup, arrangements, indentation, removing some comments etc.

milansky commented:
Controls can be (maybe even severely) improved with Pass by Reference to modify values, it would be much more cleaner but not sure about the performance.

I just realized how easily you can change variables or any value outside of the function call and now I understand why Nico was constantly saying "Constant reference!!" 

Cool drinking game, when Nico says constant reference you take a shot :D

petrSchreiber reviewed (approved):
Rated E for EPIC
